src/ray/gcs/redis_context.cc

```diff
@@ -284,21 +284,35 @@ void SetDisconnectCallback(RedisAsyncContext *redis_async_context) {
                                   RedisAsyncContextDisconnectCallback);
 }
 
+template <typename RedisContext, typename RedisConnectFunction>
+Status ConnectWithoutRetries(const std::string &address, int port,
+                             const RedisConnectFunction &connect_function,
+                             RedisContext **context, std::string &errorMessage) {
+  *context = connect_function(address.c_str(), port);
+  if (*context == nullptr || (*context)->err) {
+    std::ostringstream oss(errorMessage);
+    if (*context == nullptr) {
+      oss << "Could not allocate Redis context.";
+    } else if ((*context)->err) {
+      oss << "Could not establish connection to Redis " << address << ":" << port
+          << " (context.err = " << (*context)->err << ")";
+    }
+    return Status::RedisError(errorMessage);
+  }
+  return Status::OK();
+}
+
 template <typename RedisContext, typename RedisConnectFunction>
 Status ConnectWithRetries(const std::string &address, int port,
                           const RedisConnectFunction &connect_function,
                           RedisContext **context) {
   int connection_attempts = 0;
-  *context = connect_function(address.c_str(), port);
-  while (*context == nullptr || (*context)->err) {
+  std::string errorMessage = "";
+  Status status =
+      ConnectWithoutRetries(address, port, connect_function, context, errorMessage);
+  while (!status.ok()) {
     if (connection_attempts >= RayConfig::instance().redis_db_connect_retries()) {
-      if (*context == nullptr) {
-        RAY_LOG(FATAL) << "Could not allocate redis context.";
-      }
-      if ((*context)->err) {
-        RAY_LOG(FATAL) << "Could not establish connection to redis " << address << ":"
-                       << port << " (context.err = " << (*context)->err << ")";
-      }
+      RAY_LOG(FATAL) << errorMessage;
```

Comment:
Can you update the log message to indicate we exceed max num retries? For example, 

"Couldn't connect to Redis with within the max count " << redis_db_connect_retries << ". Error message: " << errorMessage

Reply:
Uh, sure. Like I said in https://github.com/ray-project/ray/pull/12026#issuecomment-728535839, I was specifically trying to avoid changing the error message at the same time as moving around the lines of code, but if you actually want me to, I'm happy to.

Reply:
@rkooo567 Like this? https://github.com/ray-project/ray/pull/12016/files#diff-478a807e3003772df7d270ff7ba820665462d5c82d008dff07dc0f3d4747af92R318